src/builder.rs

```diff
@@ -347,14 +354,18 @@ impl DbcBuilder {
     /// build spent proofs from shares.
     pub fn spent_proofs(&self) -> Result<BTreeSet<SpentProof>> {
         let transaction_hash = Hash::from(self.transaction.hash());
-        let spent_proofs: BTreeSet<SpentProof> = self
+        let mut spent_proofs: BTreeSet<SpentProof> = self
             .spent_proof_shares
             .iter()
             .map(|(key_image, shares)| {
                 SpentProof::try_from_proof_shares(*key_image, transaction_hash, shares.iter())
             })
             .collect::<Result<_>>()?;
 
+        self.spent_proofs.iter().for_each(|s| {
+            let _ = spent_proofs.insert(s.clone());
+        });
+
```

Comment:
What do you think of using `chain` here instead of mutating?

```suggestion
        let spent_proofs: BTreeSet<SpentProof> = self
            .spent_proof_shares
            .iter()
            .map(|(key_image, shares)| {
                SpentProof::try_from_proof_shares(*key_image, transaction_hash, shares.iter())
            })
            .chain(self.spent_proofs.iter().cloned().map(|s| Ok(s)))
            .collect::<Result<_>>()?;

```